nomic/utils.py

```diff
@@ -241,10 +242,29 @@ def get_object_size_in_bytes(obj):
 
 # Helpful function for downloading feather files
 # Best for small feather files
-def download_feather(url: str, path: Path, headers: Optional[dict] = None):
-    data = requests.get(url, headers=headers)
-    readable = BytesIO(data.content)
-    readable.seek(0)
-    tb = pa.feather.read_table(readable, memory_map=True)  # type: ignore
-    path.parent.mkdir(parents=True, exist_ok=True)
-    pa.feather.write_feather(tb, path)  # type: ignore
+def download_feather(url: str, path: Path, headers: Optional[dict] = None, retries=3, overwrite=False) -> pa.Schema:
+    """
+    Download a feather file from a URL to a local path.
+    Returns the schema of feather file if successful.
+    """
+    assert retries > 0, "Retries must be greater than 0"
+    download_attempt = 0
+    download_success = False
+    schema = None
+    while download_attempt < retries and not download_success:
+        download_attempt += 1
+        if not path.exists() or overwrite:
+            data = requests.get(url, headers=headers)
+            readable = BytesIO(data.content)
```

Comment:
What is the largest file we expect to pull down here?

Reply:
according to @bmschmidt, several hundreds of MB at max. Is this size reasonable?

Reply:
mainly we have to buffer the whole file in memory (twice) working this way - a couple hundred MB is _probably_ okay

Reply:
I think using `pa.BufferReader` instead of `BytesIO` here can avoid the second copy (`read_table` has to use the file-like-object interface to read a BytesIO and therefore has to copy it, but a `BufferReader` can be handled zero-copy)

Reply:
No: the dance is to save user hard-drive space. 

Javascript doesn't support record batch compression. So our wire format is feather with no internal record-batch compression, wrapped in `Content-Encoding: gzip` so it gets quietly compressed/decompressed by the browser or, here, `requests` or whatever.

`pa.feather.write_feather` here will write it with `snappy` or something.

Reply:
We could probably kick the can down the road and write the bytes directly with the filename "z/x/y.embeddings.feather.gz" and do `feather.read_table` on `gzip.open` around the file, but that tends to be a lot slower and preclude mem-mapping them.